pkg/kubelet/cm/dra/manager.go

```diff
@@ -98,7 +98,10 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string, n
 }
 
 func (m *ManagerImpl) GetWatcherHandler() cache.PluginHandler {
-	return cache.PluginHandler(dra.NewRegistrationHandler(m.kubeClient, m.getNode))
+	// The time that DRA drivers have to come back after being unregistered
+	// before the kubelet removes their ResourceSlices.
+	wipingDelay := 5 * time.Minute
```

Comment:
After I [finally understood](https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1930628588) that a seamless upgrade of a DaemonSet always includes some downtime, I focused on minimizing the impact.

Now ResourceSlices are preserved if the upgrade finishes within 5 minutes. The downside is that it takes 5 minutes longer in case of a permanent driver uninstall to prevent scheduling pods which depend on the driver.

Which one is more important?

cc @klueska

Reply:
> After I https://github.com/kubernetes/kubernetes/pull/129832#discussion_r1930628588 that a seamless upgrade of a DaemonSet always includes some downtime, I focused on minimizing the impact.

Wrong - if `maxSurge > 0`, two pods may run at the same time.

This change here may still be useful for the traditional deployment with `maxSurge = 0` (the default).

Reply:
Why it's 5 minutes? Can it be shorter?

Reply:
5 minutes is what most E2E tests use as "time to wait for pod start". We can make it longer or shorter.

Reply:
I'd vote for making it shorter to detect permanent plugin failures faster. How about 3 min?

Reply:
Added the pros and cons for choosing the duration in the comment for it.

Reply:
We can make it shorter. But personally I am leaning towards not doing this at all because all choices have drawbacks. Instead, DRA drivers should use seamless upgrade if they care about not dropping ResourceSlices during an upgrade, which of course hinges on actually supporting it.

Reply:
I was mostly concerned with permanent plugin failures, not with upgrades.
If 5 minutes is what we use here, I should probably increase delays/intervals in https://github.com/kubernetes/kubernetes/pull/129799 as well. WDYT?

Reply:
That seems different. Here we need a delay that covers the duration between "one pod stops normally, other pod starts". The kubelet detects the pod exit because the socket gets removed.

Your PR is about scenarios where a running pod is just unresponsive (which shouldn't last 5 minutes) vs. gone forever after an unclean shutdown. Anyway, explain in your PR how the constants were chosen and what the pros and cons of making them longer or shorter are and we can discuss there.

Reply:
> I was mostly concerned with permanent plugin failures, not with upgrades.

But we need to consider both. What is good for one (short delay after failure) is bad for the other (not enough time to finish the upgrade, which may include image pulling).

Reply:
> We can make it shorter. But personally I am leaning towards not doing this at all because all choices have drawbacks. Instead, DRA drivers should use seamless upgrade if they care about not dropping ResourceSlices during an upgrade, which of course hinges on actually supporting it.

Does "not doing this at all" mean unregistering results in immediate deletion? Would that happen during some sort of intermittent failure, or just if there was an orderly destruction and recreation? I think a bit of tolerance may be helpful. I agree with the general statement that if they want seamless upgrade, they should use `maxSurge > 0`. But if a user, say, deletes the pod, will it unregister and then reregister on recreation? Do we want resource slices to go away in that case?

Reply:
> But if a user, say, deletes the pod, will it unregister and then reregister on recreation?

Yes. Slices then get deleted on unregistration, immediately. That is the current behavior, the rationale being that we want to stop scheduling pods as soon as possible.

> I think a bit of tolerance may be helpful.

How large would you make that tolerance?

Reply:
I was thinking something short, just to prevent flapping during intermittent failures or crashes. Would 30s make a difference? A minute?

Reply:
For a container restart, 30 second might help. I'm not sure how much it would help for a real driver update, but that was the case that I would like to see handled via `maxSurge` anyway. Let's go for 30 seconds.

Reply:
I switched to 30 seconds.